src/conversions/num_bigint.rs

```diff
@@ -138,53 +140,69 @@ macro_rules! bigint_conversion {
                 unsafe {
                     let num: Py<PyLong> =
                         Py::from_owned_ptr_or_err(py, ffi::PyNumber_Index(ob.as_ptr()))?;
-
-                    let n_bytes = {
+                    let n_bits = {
                         cfg_if::cfg_if! {
                             if #[cfg(not(Py_LIMITED_API))] {
                                 // fast path
                                 let n_bits = ffi::_PyLong_NumBits(num.as_ptr());
                                 if n_bits == (-1isize as usize) {
                                     return Err(crate::PyErr::fetch(py));
-                                } else if n_bits == 0 {
-                                    0
-                                } else {
-                                    (n_bits - 1 + $is_signed) / 8 + 1
                                 }
+                                n_bits
                             } else {
                                 // slow path
                                 let n_bits_obj = num.getattr(py, crate::intern!(py, "bit_length"))?.call0(py)?;
                                 let n_bits_int: &PyLong = n_bits_obj.downcast_unchecked(py);
-                                let n_bits = n_bits_int.extract::<usize>()?;
-                                if n_bits == 0 {
-                                    0
-                                } else {
-                                    (n_bits - 1 + $is_signed) / 8 + 1
-                                }
+                                n_bits_int.extract::<usize>()?
                             }
                         }
                     };
 
-                    if n_bytes <= 128 {
-                        let mut buffer = [0; 128];
-                        extract(num.as_ref(py), &mut buffer[..n_bytes], $is_signed)?;
-                        Ok($from_bytes(&buffer[..n_bytes]))
-                    } else {
-                        let mut buffer = vec![0; n_bytes];
-                        extract(num.as_ref(py), &mut buffer, $is_signed)?;
-                        Ok($from_bytes(&buffer))
+                    if n_bits == 0 {
+                        return Ok(0.try_into().unwrap())
                     }
+
+                    let n_digits = (n_bits + 31 + $is_signed) / 32;
+
+                    let mut buffer = Vec::<u32>::with_capacity(n_digits);
+                    buffer.set_len(n_digits);
```

Comment:
I think this basically instant UB because it assumes initialization for data that is clear uninitialized.

Either we need to use raw pointers everywhere or something like [`spare_capacity_mut`](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.spare_capacity_mut) if our MSRV allows it. In any case, we cannot call `set_len` until the buffer has actually been initialized.

Reply:
Ah, that's a very good catch 👍

Reply:
We could probably call `set_len` _after_ extract, and just pass a raw pointer to `extract`

Reply:
im beyond puzzled. I Changed `buffer.len()` into a `length` parameter to the function, and when i use the function
```
let mut buffer = Vec::<u32>::with_capacity(n_digits);
                    extract(
                        num.as_ref(py),
                        &mut buffer,
                        n_digits,
                        $is_signed
                    )?;
                    buffer.set_len(n_digits);
```
it just produces wrong results. If i `set_len` before the call to extract in the abi3 mode, it works fine. If i zero-init the vector, it works fine. But if i leave the data uninit and dont set the length (bot of those things!), it produces noncence. Am i missing something here?

Reply:
In `abi3` mode there's a call to `buffer.copy_from_slice` so unless you changed that I would expect that to do nothing to a zero-length vector.

You could just sidestep the problem and call `BigInt::from_signed_bytes_le` (or similar for `BigUint`) on the bytes slice from the Python `to_bytes` call, though that will require a bit more refactoring.

Reply:
> Am i missing something here?

Difficult to tell without looking at the code, but I suspect you still take `&mut [u32]`, but this will be a zero-length slice if you do not zero-initialize the buffer. As suggested, you should take a raw pointer instead. (A slice `&mut [u32]` also _requires_ the corresponding memory to be initialized to a valid value.) 

(Note that in https://github.com/PyO3/pyo3/discussions/3365#discussioncomment-6630845 I mentioned that skipping zero-initialization would make this messy. This is what I was talking about.)

Reply:
thank you all, your crystall ball omniscience is impressive!